### PR TITLE
Multicommodity simulator storage

### PR DESCRIPTION
**Pull request by FJanssen-TNO**
This PR includes updates and tests to use storages both electricity and gas in the multicommodity simulator
To be merged after the merge of multicommodity-simulator.

### CODE REVIEW

**Comment by FJanssen-TNO** [comment]
@KobusVanRooyen ready for review

File: src/mesido/gas_physics_mixin.py [hunk 1]
```diff
@@ -573,7 +576,7 @@ def __gas_storage_discharge_path_constraints(self):
         options = self.energy_system_options()
 
         if options["gas_storage_discharge_variables"]:
-            for storage in self.energy_system_components.get("gas_storage"):
+            for storage in self.energy_system_components.get("gas_tank_storage", []):
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/mesido/gas_physics_mixin.py:579]: currently we only have a group of asset "gas_tank_storage", maybe at some point we'll expand and make gas_tank_storage a subgroup of gas_storage then the map without the tank makes sense.

File: src/mesido/workflows/multicommodity_simulator_workflow.py [hunk 2]
```diff
@@ -130,8 +131,8 @@ class MinimizeStorageGoalMerit(Goal):
     """
 
     def __init__(self, source_variable, prod_priority, func_range_bound, nominal, order=2):
-        self.target_min = 0.0  # func_range_bound[0]
-        self.target_max = func_range_bound[1] * 0.999
+        # self.target_min = 0.0  # func_range_bound[0]
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/mesido/workflows/multicommodity_simulator_workflow.py:134]: bug from my side, I hadn't properly looked at the storage goals yet. yes this comment I'll delete

File: src/mesido/gas_physics_mixin.py [hunk 3]
```diff
@@ -560,7 +563,7 @@ def __flow_direction_path_constraints(self, ensemble_member):
 
         return constraints
 
-    def __gas_storage_discharge_path_constraints(self):
+    def __gas_storage_discharge_path_constraints(self, ensemble_member):
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/mesido/gas_physics_mixin.py:566]: Thanks, forgot about the note, have deleted.

File: tests/test_multicommodity_simulator.py [hunk 4]
```diff
@@ -337,6 +341,246 @@ def energy_system_options(self):
...
+
+        check_electrolyzer_efficiency(tol, electrolyzer_gas, electrolyzer_power, esdl_electrolyzer)
+
+        # gas mass balance including storage
+        np.testing.assert_allclose(
+            electrolyzer_gas - storage_gas_mass_flow, demand_gas, rtol=tol, atol=tol
+        )
+        # timestep 7 & 8 electrolyzer should be producing more hydrogen then the demand can take.
+        # gas_demand is limited by pipe size.
+        # storage is not charging if gas demand is not maximisedd at bound
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply tests/test_multicommodity_simulator.py:393]: changed

File: tests/test_multicommodity_simulator.py [hunk 5]
```diff
@@ -337,6 +341,246 @@ def energy_system_options(self):
...
+        np.testing.assert_array_less(storage_gas_mass_flow[storage_not_charging], 0.0 + tol)
+        storage_charging = electrolyzer_gas >= gas_demand_bound[1]
+        np.testing.assert_array_less(0.0 - tol, storage_gas_mass_flow[storage_charging])
+
+    def test_multi_commodity_simulator_emerge_battery(self):
+        """
+        Test to run the multicommodity simulator including a battery and gas storage.
+        Checks:
+        - feasibility
+        - power conservation
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply tests/test_multicommodity_simulator.py:406]: updated

File: tests/test_multicommodity_simulator.py [hunk 6]
```diff
@@ -337,6 +341,246 @@ def energy_system_options(self):
...
+            esdl_parser=ESDLFileParser,
+            profile_reader=ProfileReaderFromFile,
+            input_timeseries_file="timeseries_short.csv",
+        )
+
+        results = solution.extract_results()
+        parameters = solution.parameters(0)
+
+        feasibility_test(solution)
+        # electric_power_conservation_test(solution, results)
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply tests/test_multicommodity_simulator.py:490]: accident, restored.

File: tests/test_multicommodity_simulator.py [hunk 7]
```diff
@@ -337,6 +341,246 @@ def energy_system_options(self):
...
+                    b = dw_headloss_min - a * velocities[line_num - 1]
+                    headloss_calc = a * v + b
+                    np.testing.assert_allclose(abs(head_loss[i]), headloss_calc, 0.1)
+                # else:
+                #     np.testing.assert_allclose(abs(head_loss[i]), 0.0)
+
+            # wrong line segment seems active
+            results[
+                f"{pipe}__pipe_linear_line_segment_num_3_pos_discharge"
+            ]  # instead of segment 1 since that one is upto 3m/s
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply tests/test_multicommodity_simulator.py:580]: removed

File: src/mesido/workflows/multicommodity_simulator_workflow.py [hunk 8]
```diff
@@ -506,6 +525,15 @@ def __merit_controls(self, assets_list):
                     attributes["merit_order"].append(
                         a.attributes["costInformation"].marginalCosts.value
                     )
+                    if (
+                        a.asset_type == "Electrolyzer"
+                    ):  # electrolyzer would require minimisation of electricity right after
+                        # maximisation gas
+                        attributes["asset_name"].append(f"{a.name}_prod")
+                        attributes["merit_order"].append(
+                            a.attributes["costInformation"].marginalCosts.value - 1e-6
```
[the remark this answers is not in the document]
Comment (KobusVanRooyen) [reply src/mesido/workflows/multicommodity_simulator_workflow.py:535]: Comment has been added to make it more clear why the value is used

**Review by KobusVanRooyen — APPROVED** [review]

**Pull request merged**